pkg/agent/sysadvisor/plugin/inference/modelresultfetcher/borwein/borwein.go

```diff
@@ -0,0 +1,328 @@
+	}
+
+	req.FeatureNames = append(req.FeatureNames, bmrf.nodeFeatureNames...)
+	req.FeatureNames = append(req.FeatureNames, bmrf.containerFeatureNames...)
+
+	nodeFeatureValues := make([]string, 0, len(bmrf.nodeFeatureNames))
+	for _, nodeFeatureName := range bmrf.nodeFeatureNames {
+		if GetNodeFeatureValue == nil {
+			return nil, fmt.Errorf("nil GetNodeFeatureValue")
+		}
+
+		nodeFeatureValue, err := GetNodeFeatureValue(nodeFeatureName, metaServer, metaReader)
+
+		if err != nil {
+			return nil, fmt.Errorf("get node feature: %v failed with error: %v", nodeFeatureName, err)
+		}
+
+		nodeFeatureValues = append(nodeFeatureValues, nodeFeatureValue)
+	}
+
+	for _, pod := range pods {
+		if pod == nil {
+			general.Warningf("nil pod")
+			continue
+		}
+
+		req.PodRequestEntries[string(pod.UID)] = &borweininfsvc.ContainerRequestEntries{
+			ContainerFeatureValues: make(map[string]*borweininfsvc.FeatureValues, len(pod.Spec.Containers)),
+		}
+
+		foundMainContainer := false
+		for i := 0; i < len(pod.Spec.Containers); i++ {
+			containerName := pod.Spec.Containers[i].Name
+			containerInfo, ok := metaReader.GetContainerInfo(string(pod.UID), containerName)
+
+			if !ok {
+				return nil, fmt.Errorf("GetContainerInfo for pod: %s/%s, container: %s failed",
+					pod.Namespace, pod.Name, containerName)
+			}
+
+			if containerInfo.ContainerType == v1alpha1.ContainerType_MAIN {
+
+				foundMainContainer = true
+				unionFeatureValues := &borweininfsvc.FeatureValues{
+					Values: make([]string, 0, len(req.FeatureNames)),
+				}
+
+				unionFeatureValues.Values = append(unionFeatureValues.Values, nodeFeatureValues...)
+
+				for _, containerFeatureName := range bmrf.containerFeatureNames {
+					if GetContainerFeatureValue == nil {
+						return nil, fmt.Errorf("nil GetContainerFeatureValue")
+					}
+
+					containerFeatureValue, err := GetContainerFeatureValue(string(pod.UID),
+						containerName, containerFeatureName,
+						metaServer, metaReader)
+
+					if err != nil {
+						return nil, fmt.Errorf("GetContainerFeatureValue for pod: %s/%s, container: %s failed",
+							pod.Namespace, pod.Name, containerName)
+					}
+
+					unionFeatureValues.Values = append(unionFeatureValues.Values, containerFeatureValue)
+				}
+
+				req.PodRequestEntries[string(pod.UID)].ContainerFeatureValues[containerName] = unionFeatureValues
+				break
+			}
+		}
+
+		// todo: currently only inference for main container,
+		// maybe supporting sidecar later
+		if !foundMainContainer {
+			return nil, fmt.Errorf("main container isn't found for pod: %s/%s", pod.Namespace, pod.Name)
+		}
+	}
+
+	return req, nil
+}
+
+// initAdvisorClientConn initializes memory-advisor related connections
+func (bmrf *BorweinModelResultFetcher) initInferenceSvcClientConn() error {
+	if bmrf.inferenceServiceSocketAbsPath == "" {
+		return fmt.Errorf("empty inferenceServiceSocketAbsPath")
+	}
+
+	infSvcConn, err := process.Dial(bmrf.inferenceServiceSocketAbsPath, 5*time.Second)
```

Comment:
to validate if it's ok when inference server starts after BorweinModelResultFetcher